File: DataStructures/UnionFind/QuickFind.java
package DataStructures.UnionFind;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class QuickFind {
    /*
    * Maintain an integer array of size N where N == number of objects.
    * Index == Object id
    * Array entry == Root
    * If two objects have the same root, they are connected.
    */
    private int[] root;

    /*
    * Constructor
    * O(N)
    */
    public QuickFind(int size) {
        root = new int[size];
        for (int i = 0; i < size; i++) { // Set each object's root to itself
            root[i] = i;
        }
    }

    /*
    * Check if objects are connected
    * O(1)
    */
    public boolean connected(int p, int q) {
        return root[p] == root[q];
    }

    /*
    * To connect two objects, we change all entries of root[q] to equal root[p]
    * O(N)
    */
    public void union(int p, int q) {
        if (!connected(p, q)) {
            int pRoot = root[p];
            int qRoot = root[q];
            for (int i = 0; i < root.length; i++) {
                if (root[i] == qRoot) {
                    root[i] = pRoot;
                }
            }
        }
    }

    /*
     * Test Client: 
     * - Read in pairs of integers from standard input. 
     * - If they are not yet connected, connect them and print out pair
     */
    public static void main(String args[]) throws IOException {
        File testFile = new File(System.getProperty("user.dir") + "/Tests/UF10.txt");
        BufferedReader br = new BufferedReader(new FileReader(testFile));
        int size = Integer.parseInt(br.readLine());
        QuickFind qf = new QuickFind(size);
        String pair = br.readLine();
        while (pair != null) {
            int p = Integer.parseInt(pair.substring(0, 1));
            int q = Integer.parseInt(pair.substring(2));
            if (qf.connected(p, q)) {
                System.out.println("[" + p + ", " + q + "] " + "are already connected.");
            } else {
                qf.union(p, q);
                System.out.println("[" + p + ", " + q + "] " + "are now connected.");
            }
            pair = br.readLine();
        }
        br.close();
    }
}

File: DataStructures/UnionFind/QuickFind.js
const fs = require('fs')

class QuickFind {
    /*
    * Maintain an integer array of size N where N == number of objects.
    * Index == Object id | Array entry == Root
    * If two objects have the same root, they are connected.
    */
    constructor(size) {
        this.roots = new Array(size)
        for (let i = 0; i < size; i++) { // Set each object's root to itself
            this.roots[i] = i
        }
    }

    /*
    * Check if two objects are connected
    * O(1)
    */
    connected(p, q) {
        return this.roots[p] === this.roots[q]
    }

    /*
    * To connect two objects, we change all entries of root[q] to equal root[p]
    * O(N)
    */
   union(p, q) {
       if (!this.connected(p, q)) {
           const pRoot = this.roots[p]
           const qRoot = this.roots[q]
           for (let i = 0; i < this.roots.length; i++) {
               if (this.roots[i] == qRoot) {
                   this.roots[i] = pRoot
               }
           }
       }
   }
}

/*
*   Test Section: 
* - Read in pairs of integers from standard input. 
* - If they are not yet connected, connect them and print out pair
*/
function main() {
    fs.readFile(__dirname + '/Tests/UF10.txt', 'utf8', 
    function(err, data) {
        if (err) throw err
        const commands = data.split('\n')
        const quickFind = new QuickFind(Number(commands[0]))
        for (let i = 1; i < commands.length; i++) {
            const p = Number(commands[i][0])
            const q = Number(commands[i][2]) 
            if (quickFind.connected(p, q)) {
                console.log(`[${p}, ${q}] are already connected.`)
            } else {
                quickFind.union(p, q)
                console.log(`[${p}, ${q}] are now connected.`)
            }
        }
    })
}

main()
